### ISSUE TITLE
How does DataLoaderDocumentListener work as a Singleton?

### ISSUE THREAD

**Issue by GrangerAts** [graphql-dotnet/graphql-dotnet#1192]
How is this even working for anyone? I can't see how the implementation for DataLoaderDocumentListener in 'master' could ever work when scoped as Singleton.

So let's start with Startup in the [sample server in graphql-dotnet/server](https://github.com/graphql-dotnet/server/blob/master/samples/Samples.Server/Startup.cs). 

First, DI services get configured...
1. The `AddGraphQL()` extension method does this:  
```
services.TryAddSingleton<IDocumentExecuter, DocumentExecuter>();
services.AddTransient(typeof(IGraphQLExecuter<>), typeof(DefaultGraphQLExecuter<>));
```
2. Then the `AddDataLoader()` extension method does this:
```
builder.Services.TryAddSingleton<IDataLoaderContextAccessor, DataLoaderContextAccessor>();
builder.Services.AddSingleton<IDocumentExecutionListener, DataLoaderDocumentListener>();
```

Then the request pipeline gets configured by using the `UseGraphQL()` extension method. That method uses the `GraphQLHttpMiddleware` class to setup how the request is handled. And ultimately, that uses the `DefaultGraphQLExecuter` class which, uses the _Singleton_ `DocumentExecutor ` from the DI container. The `DefaultGraphQLExecuter` also gets all registered `IDocumentExecutionListener` objects injected into it when it's created; that's how it gets the `DataLoaderDocumentListener` registered during the `AddDataLoader()` call.

When a `DataLoaderDocumentListener` is created, an `IDataLoaderContextAccessor` is injected (also a Singleton). The `DataLoaderContextAccessor`'s `Context` is just a spot to hold a reference to a `DataLoaderContext`; it doesn't get setup by the object itself, but relies on code that uses it to manage the property. This "Context" property is the one people are complaining about being _null_ in issue #866. It's the one people use to call `GetOrAddBatchLoader()`, for example.

So now we're to my confusion.

1. The Singleton `DataLoaderDocumentListener` is used on each request, and each request raises that object's events.
2. The `BeforeExecutionAsync` event does this:
```
        public Task BeforeExecutionAsync(object userContext, CancellationToken token)
        {
            if (_accessor.Context == null)
                _accessor.Context = new DataLoaderContext();

            return Task.CompletedTask;
       }
```
3. The `AfterExecutionAsync` event does this:
```
        public Task AfterExecutionAsync(object userContext, CancellationToken token)
        {
            _accessor.Context = null;

            return Task.CompletedTask;
        }
```

It wouldn't matter if these used a thread `lock`; it still shouldn't work in a multi-threaded situation. 

1. A thread/request comes in and gets suspended in the middle of execution. 
1. A second thread/request comes in and executes to completion. 
1. The first thread resumes and attempts to access `.Context`. It's going to get a `NullReferenceException `because the earlier thread null'd it out when it finished executing its query. 

Shouldn't `DataLoaderDocumentListener` and `DataLoaderContextAccessor` be Scoped?

Or is this particular issue not biting anyone because it's some artifact from earlier development and no one actually uses _that_ particular "Context"? 
...because it occurs to me that all of the `ObjectGraphType` objects in a Schema's Query get instantiated by something inside the GraphQl.Net system (the Schema class?). And that would mean that the object that's creating the DataLoaderContext object for an injected IDataLoaderContextAccessor would be in control, not the `DataLoaderDocumentListener`. If it's using the .Net core DI container, then there's still a problem, since it's Singleton. But if it makes its own, then it's just a hidden dependency. 

Did I mention that I'm really, really confused? :)

**Comment by GrangerAts** [comment]
So one more speculation: 
Perhaps the fact that IDataLoaderContextAccessor.Context is an `AsyncLocal<DataLoaderContext>` under the hood is the crux? I certainly don't understand that paradigm fully. From what I think I know:

1. `AsyncLocal<>` stores its data scoped to the current thread. So that would mean that even though the `DataLoaderContextAccessor` is Singleton, the `Context` property always stores data as if it were Scoped (well, depending on the framework Gql.Net is running within).
1. If people follow the recommendation to use **ConfigureAwait(false)** (e.g.  `ExecutionResult result = await _documentExecuter.ExecuteAsync(executionOptions).ConfigureAwait(false);` ), then after an await resumes, `Context` will return a different object, or null, because the continuation is executing on a different thread.

If that's correct, then that could explain some observed behavior in #866 as well.

**Comment by johnrutherford** [comment]
1. Yes, it works because the `DataLoaderContextAccessor.Context` property is `AsyncLocal<>`.

2. Correct. For v3.0, I plan on refactoring to store the `DataLoaderContext` on the `UserContext` dictionary. This will make things simpler and fix issues with the async context changing.

**Comment by sungam3r** [comment]
2. No, `AsyncLocal<T>` does not affected by `ConfigureAwait(false)`. `ConfigureAwait(false)` is about not to flow `SynchronizationContext`, but not about logical call context. You can’t “opt-out” of the logical call context. If it was not saved, then almost all of the .NET code would fall apart.

**Comment by GrangerAts** [comment]
@sungam3r  - OK, so I wasn't sure from what you wrote, but it all looks good to me now. The "execution context" is the "logical call context" in the async world; people don't use "thread local storage" anymore. [1](https://devblogs.microsoft.com/pfxteam/executioncontext-vs-synchronizationcontext/), [2](https://blog.stephencleary.com/2013/04/implicit-async-context-asynclocal.html).  

`AsyncLocal<T>` is not affected by `ConfigureAwait(false)` because it stores its data with the execution context, not the thread.

`AsyncLocal<T>` internally uses `ExecutionContext.GetLocalValue(this);`  and `ExecutionContext.SetLocalValue(this, value, m_valueChangedHandler != null);`. Both of those internally use `Thread.CurrentThread._executionContext`  to get/set data for the `AsyncLocal<T>` instance. ( [3](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Threading/AsyncLocal.cs), [4](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs) )

So, in summary, aside from me being confused trying to follow things in the code, how a Singleton effectively acts as Scoped, it's working OK as-is.
